G3-G4 wrap-up (#8641)
Mbed-TLS/mbedtls: docs/driver-only-builds.md

`[PSA_WANT|MBEDTLS_PSA_ACCEL]_ALG_FFDH` allow builds accelerating FFDH and
removing builtin support (i.e. `MBEDTLS_DHM_C`).

Ciphers and AEADs
-----------------

> It is possible to have all ciphers and AEADs operations provided only by a

Reviewer: ```suggestion
It is possible to have all ciphers and AEAD operations provided only by a
```

driver. More precisely, for each desired combination of key type and
algorithm/mode you can:
- enable desired PSA key type(s):
  - `PSA_WANT_KEY_TYPE_AES`,
  - `PSA_WANT_KEY_TYPE_ARIA`,
  - `PSA_WANT_KEY_TYPE_CAMELLIA`,
  - `PSA_WANT_KEY_TYPE_CHACHA20`,

>   - `PSA_WANT_KEY_TYPE_DES`;

Reviewer: ```suggestion
  - `PSA_WANT_KEY_TYPE_DES`.
```

- enable desired PSA algorithm(s):
  - unauthenticated ciphers modes:
    - `PSA_WANT_ALG_CBC_NO_PADDING`,
    - `PSA_WANT_ALG_CBC_PKCS7`,
    - `PSA_WANT_ALG_CCM_STAR_NO_TAG`,
    - `PSA_WANT_ALG_CFB`,
    - `PSA_WANT_ALG_CTR`,
    - `PSA_WANT_ALG_ECB_NO_PADDING`,
    - `PSA_WANT_ALG_OFB`,

>     - `PSA_WANT_ALG_STREAM_CIPHER`;

Reviewer: ```suggestion
    - `PSA_WANT_ALG_STREAM_CIPHER`.
```

  - AEADs:
    - `PSA_WANT_ALG_CCM`,
    - `PSA_WANT_ALG_GCM`,

>     - `PSA_WANT_ALG_CHACHA20_POLY1305`;

Reviewer: ```suggestion
    - `PSA_WANT_ALG_CHACHA20_POLY1305`.
```

- enable `MBEDTLS_PSA_ACCEL_[KEY_TYPE_xxx|ALG_yyy]` symbol(s) which correspond

>    to the PSA_WANT_KEY_TYPE_xxx` and `PSA_WANT_ALG_yyy` of the previous steps;

Reviewer: ```suggestion
   to the PSA_WANT_KEY_TYPE_xxx` and `PSA_WANT_ALG_yyy` of the previous steps.
```

- disable builtin support of key types:
  - `MBEDTLS_AES_C`,
  - `MBEDTLS_ARIA_C`,
  - `MBEDTLS_CAMELLIA_C`,
  - `MBEDTLS_DES_C`,

>   - `MBEDTLS_CHACHA20_C`;

Reviewer: ```suggestion
  - `MBEDTLS_CHACHA20_C`.
```

  and algorithms/modes:

>   - `MBEDTLS_CBC_C`

Reviewer: Seems a bit inconsistent to not use commas and full stops like you have done above.

  - `MBEDTLS_CFB_C`
  - `MBEDTLS_CTR_C`
  - `MBEDTLS_OFB_C`
  - `MBEDTLS_XTS_C`
  - `MBEDTLS_CCM_C`
  - `MBEDTLS_GCM_C`
  - `MBEDTLS_CHACHAPOLY_C`
  - `MBEDTLS_NULL_CIPHER`

> Once a key type and related algorithm are accelerated, all cipher and AEADs

Reviewer: ```suggestion
Once a key type and related algorithm are accelerated, all cipher and AEAD
```

Author: This part has been changed by a following commit.

operations of that type requested through the PSA Crypto API are performed by

> the driver. Only functions belonging to legacy modules which are disabled won't

Reviewer 2: The last sentence is not correct. Functions like `mbedtls_pkcs5_pbes2_ext()` are not going to be available either. Also, some functions are available but with a reduced feature set.

I think we should replace this sentence with "All the PSA Crypto APIs will work, as well as X.509 and TLS (with `MBEDTLS_USE_PSA_CRYPTO` enabled) but some non-PSA APIs will be absent or have reduced functionality, see the next section for details".

In order to remember the full list of details, we can have a look at ignored things for `analyze_driver_vs_reference_cipher_aead` in `analyze_outcomes.py` plus things that are disabled in the reference component. So, if I'm not missing anything:
- PKCS5/PKCS12 en/decryption won't work (key derivations function from those modules still work).
- Parsing encrypted PEM won't work.
- Parsing of encrypted keys (PEM or PKCS12) in PK parse won't work.
- CMAC will not be able to use AES if `AES_C` is disabled (could still work with DES)*.
- NIST-KW needs to be disabled if `AES_C` is disabled.

Actually looking again at those test components, I'm no longer sure disabling `CIPHER_C` in the accelerated component was the right move. Perhaps we should have two accelerated components: one with `CIPHER_C` disabled and one with it enabled - to ensure that the parts of PKCS5/12 that are suppose to work indeed do.

Or perhaps instead we should just update the dependencies of PKCS5/12? We don't need to have the whole modules depending on `CIPHER_C`, but only the functions that are about encryption.

*We should not ignore the whole CMAC test suite in `analyze_driver_vs_reference_cipher_aead`, only the cases that use AES.

Author: > Or perhaps instead we should just update the dependencies of PKCS5/12? We don't need to have the whole modules depending on CIPHER_C, but only the functions that are about encryption.

Sounds good! I will add this.

> We should not ignore the whole CMAC test suite in analyze_driver_vs_reference_cipher_aead, only the cases that use AES.

Not sure about this. `analyze_driver_vs_reference_cipher_aead` uses `test_psa_crypto_config_accel_cipher_aead` as accelerated component and this, as its name suggests, accelerates *all* ciphers and AEADs. Therefore not only `AES_C` is disabled in the accelerated component, but also `DES_C`. As a consequence *all* test of `test_suite_cmac` are actually to be ignored. Am I missing something?

Reviewer 2: Nope, you not missing anything, I was - for some reason I forgot `DES_C` was disabled as well. So, please ignore that part of my comment and sorry for the noise.

Reviewer 2: Though I'll note the component's name is slightly misleading as it says `cipher_aead` but it also accelerates `cmac` which is none of these (as the name implies, it's a Message Authentication Code). Should we rename to `accel_cipher_aead_cmac`? (Makes sense as `accel` components are grouped by families of key types, and ciphers, AEADs and CMAC all use the same family of key types.)

Reviewer 2: I think it would really be clearer to rename to `accel_cipher_aead_cmac`, can you do it?

> the driver. Only functions belonging to legacy modules which are disabled won't
> be available in this configuration.

Reviewer: ```suggestion
the driver. The only functions that won't be available in this configuration are those belonging to disabled legacy modules.
```

Author: This part has been changed by a following commit.

### Restrictions

> - If an algorithm other than GCM and CCM (see

Reviewer: ```suggestion
- If an algorithm other than CCM and GCM (see
```

Reviewer: Tiny nit, I know, sorry, it's just that everywhere else CCM comes first, so I think the consistency is nice. Feel free to ignore me though.

  ["Partial acceleration for CCM/GCM"](#partial-acceleration-for-ccmgcm) below)
  is enabled but not accelerated, then all key types than can be used with it
  will need to be built-in;

> - if a key type is enabled but not accelerated, then all algorithms than can be

Reviewer: ```suggestion
- if a key type is enabled but not accelerated, then all algorithms that can be
```

  used with it will need to be built-in.

### Legacy <-> PSA matching

> It should be noticed that the matching between legacy (i.e. `MBEDTLS_xxx_C`)

Reviewer: Instead of "matching" what about "pairing" or "relationship"?

and PSA (i.e. `PSA_WANT_xxx`) symbols is not always 1:1. For example:

> - ECB mode is always enabled in legacy configuration for each key type that

Reviewer: ```suggestion
- ECB mode is always enabled in the legacy configuration for each key type that
```

  allows it (AES, ARIA, Camellia, DES), whereas it must be explicitly enabled
  in PSA with `PSA_WANT_ALG_ECB_NO_PADDING`;
- similarly for stream ciphers, it is automatically enabled for key types that

>   support it (`CHACHA20_C` and `NULL_CIPHER`) whereas it must be explicitly

Reviewer: ```suggestion
  support it (`CHACHA20_C` and `NULL_CIPHER`), whereas it must be explicitly
```

Author: This part has been changed by a following commit.

  enabled in PSA with `PSA_WANT_ALG_STREAM_CIPHER`;
- legacy symbol `MBEDTLS_CCM_C` adds support for both cipher and AEAD, whereas
  in PSA there are 2 different symbols: `PSA_WANT_ALG_CCM_STAR_NO_TAG` and
  `PSA_WANT_ALG_CCM`, respectively.

### Partial acceleration for CCM/GCM

> [This section depends on #8598 so it might updated while that PR progresses.]

Reviewer: ```suggestion
[This section depends on #8598 so it might be updated while that PR progresses.]
```

> In case legacy CCM/GCM algorithms are enabled it is still possible to benefit

Reviewer: ```suggestion
In case legacy CCM/GCM algorithms are enabled, it is still possible to benefit
```

from PSA acceleration by enabling support for ECB mode
(`PSA_WANT_ALG_ECB_NO_PADDING`) together with desired key type(s)

> (`PSA_WANT_KEY_TYPE_[AES|ARIA|CAMELLIA]`). In such configuration it is possible

Reviewer: ```suggestion
(`PSA_WANT_KEY_TYPE_[AES|ARIA|CAMELLIA]`). In such configurations it is possible
```

to:
- still benefit from legacy functions belonging to CCM/GCM modules
  (`mbedtls_[ccm|gcm]_xxx()`),

> - disable legacy key types (`MBEDTLS_[AES|ARIA|CAMELLIA]_C`) if there is no
>   other dependency requiring them, of course.

Reviewer: ```suggestion
- disable legacy key types (`MBEDTLS_[AES|ARIA|CAMELLIA]_C`), if there is no
  other dependency requiring them.
```

> ChaChaPoly has not such feature, so it requires full acceleration (key type +

Reviewer: ```suggestion
ChaChaPoly has no such feature, so it requires full acceleration (key type +
```

algorithm) in order to work with a driver.

### CTR-DRBG

Legacy CTR-DRBG module (enabled by `MBEDTLS_CTR_DRBG_C`) can also benefit from
PSA acceleration when:

> - legacy AES module is not enabled (`MBEDTLS_AES_C`) and

Reviewer: ```suggestion
- the legacy AES module is not enabled (`MBEDTLS_AES_C`) and
```

> - AES is supported on PSA side together with ECB mode, i.e.

Reviewer: ```suggestion
- AES is supported on the PSA side together with ECB mode, i.e.
```

  `PSA_WANT_KEY_TYPE_AES` + `PSA_WANT_ALG_ECB_NO_PADDING`.

### Disabling CIPHER_C

This only depends on unauthenticated ciphers: they can be either completely
accelerated or disabled in order to remove the dependency on `MBEDTLS_CIPHER_C`.

> AEADs do not have such restriction. Of course they can be accelerated as well,

Reviewer: ```suggestion
AEADs do not have such a restriction. Of course they can be accelerated as well,
```

but they can also rely on the legacy modules (`MBEDTLS_[CCM|GCM|CHACHAPOLY]`)
with the following conditions on the underlying key types:
- CCM/GCM can either use legacy key type modules `MBEDTLS_[AES|ARIA|CAMELLIA]_C`
  or their accelerated version, as described in section
  ["Partial acceleration for CCM/GCM"](#partial-acceleration-for-ccmgcm).

> - ChaChaPoly instead can only rely on legacy key type module `MBEDTLS_CHACHA20_C`

Reviewer: ```suggestion
- ChaChaPoly instead can only rely on the legacy key type module `MBEDTLS_CHACHA20_C`
```

  and algorithm `MBEDTLS_POLY1305_C`.

> It should be noticed that disabling `MBEDTLS_CIPHER_C` helps in reducing code's

Reviewer: ```suggestion
It should be noticed that disabling `MBEDTLS_CIPHER_C` helps to reduce the code's
```

`[PSA_WANT|MBEDTLS_PSA_ACCEL]_ALG_FFDH` allow builds accelerating FFDH and
removing builtin support (i.e. `MBEDTLS_DHM_C`).

Ciphers and AEADs
-----------------

> 

Reviewer 2: I believe what's available via `mbedtls_cipher_xxx` is:
- For things not using `mbedtls_cipher_setup_psa()`, exactly what has the legacy symbol enabled (`AES_C`, `CIPHER_MODE_xxx`, `GCM_C`, etc). But at the moment we don't have an actual test build for that - we'd need a test with some things accelerated and some not accelerated, because in the current test everything is accelerated so nothing works in `mbedtls_cipher_`.
- When using `mbedtls_cipher_setup_psa()` there might be a few more things working but I'm not sure. (However the function is deprecated, so perhaps not a priority.

> 

Author: So, if I understood correctly, I should wait for the new test component before modifying this section, am I right?

> 

Reviewer 2: Perhaps just leave a TODO for now, then. That is, describe everything we know about other modules, but when it comes to `mbedtls_cipher_` just say "TODO". Or we can be bold and describe what we think should work in such a build, together with a "(TODO: confirm)".

  in PSA there are 2 different symbols: `PSA_WANT_ALG_CCM_STAR_NO_TAG` and
  `PSA_WANT_ALG_CCM`, respectively.

### Partial acceleration for CCM/GCM

[This section depends on #8598 so it might updated while that PR progresses.]

In case legacy CCM/GCM algorithms are enabled it is still possible to benefit
from PSA acceleration by enabling support for ECB mode
(`PSA_WANT_ALG_ECB_NO_PADDING`) together with desired key type(s)
(`PSA_WANT_KEY_TYPE_[AES|ARIA|CAMELLIA]`). In such configuration it is possible
to:
- still benefit from legacy functions belonging to CCM/GCM modules
  (`mbedtls_[ccm|gcm]_xxx()`),
- disable legacy key types (`MBEDTLS_[AES|ARIA|CAMELLIA]_C`) if there is no
  other dependency requiring them, of course.

ChaChaPoly has not such feature, so it requires full acceleration (key type +
algorithm) in order to work with a driver.

### CTR-DRBG

Legacy CTR-DRBG module (enabled by `MBEDTLS_CTR_DRBG_C`) can also benefit from
PSA acceleration when:
- the legacy AES module is not enabled (`MBEDTLS_AES_C`) and
- AES is supported on the PSA side together with ECB mode, i.e.
  `PSA_WANT_KEY_TYPE_AES` + `PSA_WANT_ALG_ECB_NO_PADDING`.

### Disabling CIPHER_C

> This only depends on unauthenticated ciphers: they can be either completely

Reviewer 2: I think "this" refers to the section title: disabling CIPHER_C, but I agree it's not clear.

We should start with explaining the goal, for example:

----- BEGIN SUGGESTION -----
It is possible to save code size by disabling `MBEDTLS_CIPHER_C` when all of the following conditions are met:
- The application is not using the `mbedtls_cipher_` API.
- In PSA, all unauthenticated (that is, non-AEAD) ciphers are either disabled or fully accelerated (that is, all compatible key types are accelerated too).
- Either TLS is disabled, or `MBEDTLS_USE_PSA_CRYPTO` is enabled.
- `MBEDTLS_NIST_KW` is disabled.

In such a build, everything will work as usual except for the following:
- Encryption/decryption functions from the PKCS5 and PKCS12 module will not be available (only key derivation functions).
- Parsing of PKCS5- or PKCS12-encrypted keys in PK parse will fail.

Note:  AEAD ciphers (CCM, GCM, ChachaPoly) do not have a dependency on `MBEDTLS_CIPHER_C` even when using the built-in implementations.
----- END SUGGESTION -----

accelerated or disabled in order to remove the dependency on `MBEDTLS_CIPHER_C`.

AEADs do not have such a restriction. Of course they can be accelerated as well,

> but they can also rely on the legacy modules (`MBEDTLS_[CCM|GCM|CHACHAPOLY]`)

Reviewer 2: I think the whole discussion of acceleration for AEADs is off-topic here, as it's not related to disabling `CIPHER_C`.

`[PSA_WANT|MBEDTLS_PSA_ACCEL]_ALG_FFDH` allow builds accelerating FFDH and
removing builtin support (i.e. `MBEDTLS_DHM_C`).

> Ciphers and AEADs

Reviewer 2: I know I've been talking about ciphers & AEADs during the whole quarter, but that was sloppy wording: as Gilles correctly pointed out, AEADs are ciphers too. I'm suggesting we change this to "Ciphers (unauthenticated and AEAD)".

-----------------

It is possible to have all ciphers and AEAD operations provided only by a
driver. More precisely, for each desired combination of key type and
algorithm/mode you can:
- Enable desired PSA key type(s):
  - `PSA_WANT_KEY_TYPE_AES`,
  - `PSA_WANT_KEY_TYPE_ARIA`,
  - `PSA_WANT_KEY_TYPE_CAMELLIA`,
  - `PSA_WANT_KEY_TYPE_CHACHA20`,
  - `PSA_WANT_KEY_TYPE_DES`.
- Enable desired PSA algorithm(s):
  - Unauthenticated ciphers modes:
    - `PSA_WANT_ALG_CBC_NO_PADDING`,
    - `PSA_WANT_ALG_CBC_PKCS7`,
    - `PSA_WANT_ALG_CCM_STAR_NO_TAG`,
    - `PSA_WANT_ALG_CFB`,
    - `PSA_WANT_ALG_CTR`,
    - `PSA_WANT_ALG_ECB_NO_PADDING`,
    - `PSA_WANT_ALG_OFB`,
    - `PSA_WANT_ALG_STREAM_CIPHER`.
  - AEADs:
    - `PSA_WANT_ALG_CCM`,
    - `PSA_WANT_ALG_GCM`,
    - `PSA_WANT_ALG_CHACHA20_POLY1305`.
- Enable `MBEDTLS_PSA_ACCEL_[KEY_TYPE_xxx|ALG_yyy]` symbol(s) which correspond

>    to the PSA_WANT_KEY_TYPE_xxx` and `PSA_WANT_ALG_yyy` of the previous steps.

Reviewer 2: Markdown typo: missing backtick before `PSA_WANT_KEY_TYPE_xxx`.

- Disable builtin support of key types:
  - `MBEDTLS_AES_C`,
  - `MBEDTLS_ARIA_C`,
  - `MBEDTLS_CAMELLIA_C`,
  - `MBEDTLS_DES_C`,
  - `MBEDTLS_CHACHA20_C`;
  and algorithms/modes:
  - `MBEDTLS_CBC_C`
  - `MBEDTLS_CFB_C`
  - `MBEDTLS_CTR_C`
  - `MBEDTLS_OFB_C`
  - `MBEDTLS_XTS_C`
  - `MBEDTLS_CCM_C`
  - `MBEDTLS_GCM_C`
  - `MBEDTLS_CHACHAPOLY_C`
  - `MBEDTLS_NULL_CIPHER`

Once a key type and related algorithm are accelerated, all the PSA Crypto APIs

> will work, as well as X.509 and TLS (with MBEDTLS_USE_PSA_CRYPTO enabled) but

Reviewer 2: Markdown: backticks around `MBEDTLS_USE_PSA_CRYPTO`.

some non-PSA APIs will be absent or have reduced functionality, see
[Disabling CIPHER_C](#disabling-cipher_c) for details.

### Restrictions

- If an algorithm other than GCM and CCM (see
  ["Partial acceleration for CCM/GCM"](#partial-acceleration-for-ccmgcm) below)
  is enabled but not accelerated, then all key types than can be used with it
  will need to be built-in.
- If a key type is enabled but not accelerated, then all algorithms than can be
  used with it will need to be built-in.

> 

Reviewer 2: Please add "Some built-in modules are not able to take advantage of PSA drivers at the moment - TODO: list them and explain limitations."

### Legacy <-> PSA matching

Note that the matching between legacy (i.e. `MBEDTLS_xxx_C`) and PSA
(i.e. `PSA_WANT_xxx`) symbols is not always 1:1. For example:
- ECB mode is always enabled in legacy configuration for each key type that
  allows it (AES, ARIA, Camellia, DES), whereas it must be explicitly enabled
  in PSA with `PSA_WANT_ALG_ECB_NO_PADDING`.

> - In the legacy API, MBEDTLS_CHACHA20_C enables the ChaCha20 stream cipher, and

Reviewer 2: Markdown: missing backticks.

  enabling MBEDTLS_CHACHAPOLY_C also enables the ChaCha20-Poly1305 AEAD. In the
  PSA API, you need to enable PSA_KEY_TYPE_CHACHA20 for both, plus
  PSA_ALG_STREAM_CIPHER or PSA_ALG_CHACHA20_POLY1305 as desired.
- The legacy symbol `MBEDTLS_CCM_C` adds support for both cipher and AEAD,
  whereas in PSA there are 2 different symbols: `PSA_WANT_ALG_CCM_STAR_NO_TAG`
  and `PSA_WANT_ALG_CCM`, respectively.

### Partial acceleration for CCM/GCM

[This section depends on #8598 so it might updated while that PR progresses.]

In case legacy CCM/GCM algorithms are enabled it is still possible to benefit
from PSA acceleration of the underlying block cipher by enabling support for
ECB mode (`PSA_WANT_ALG_ECB_NO_PADDING`) together with desired key type(s)
(`PSA_WANT_KEY_TYPE_[AES|ARIA|CAMELLIA]`). In such configuration it is possible
to:

> - Still benefit from legacy functions belonging to CCM/GCM modules

Reviewer 2: That probably goes without saying but it doesn't hurt to be explicit about it anyway: use CCM and GCM via the PSA Crypto APIs (let's make that the first bullet point, before the others).

  (`mbedtls_[ccm|gcm]_xxx()`).
- Disable legacy key types (`MBEDTLS_[AES|ARIA|CAMELLIA]_C`) if there is no
  other dependency requiring them, of course.

ChaChaPoly has not such feature, so it requires full acceleration (key type +
algorithm) in order to work with a driver.

### CTR-DRBG

The legacy CTR-DRBG module (enabled by `MBEDTLS_CTR_DRBG_C`) can also benefit
from PSA acceleration if both of the following conditions are met:
- The legacy AES module (`MBEDTLS_AES_C`) is not enabled and
- AES is supported on the PSA side together with ECB mode, i.e.
  `PSA_WANT_KEY_TYPE_AES` + `PSA_WANT_ALG_ECB_NO_PADDING`.

> ### Disabling CIPHER_C

Reviewer 2: Minor: it's OK to use shortcuts between ourselves, but in user documentation we should use the full name: `MBEDTLS_CIPHER_C`.

It is possible to save code size by disabling MBEDTLS_CIPHER_C when all of the 
following conditions are met:
- The application is not using the `mbedtls_cipher_` API.
- In PSA, all unauthenticated (that is, non-AEAD) ciphers are either disabled or
  fully accelerated (that is, all compatible key types are accelerated too).
- Either TLS is disabled, or `MBEDTLS_USE_PSA_CRYPTO` is enabled.

> - `MBEDTLS_NIST_KW` is disabled.

Reviewer 2: Missing bullet point: "`MBEDTLS_CMAC_C` is disabled. (Note: support for CMAC in PSA can be provided by a driver.)"

  - AEADs:
    - `PSA_WANT_ALG_CCM`,
    - `PSA_WANT_ALG_GCM`,
    - `PSA_WANT_ALG_CHACHA20_POLY1305`.
- Enable `MBEDTLS_PSA_ACCEL_[KEY_TYPE_xxx|ALG_yyy]` symbol(s) which correspond
   to the `PSA_WANT_KEY_TYPE_xxx` and `PSA_WANT_ALG_yyy` of the previous steps.
- Disable builtin support of key types:
  - `MBEDTLS_AES_C`,
  - `MBEDTLS_ARIA_C`,
  - `MBEDTLS_CAMELLIA_C`,
  - `MBEDTLS_DES_C`,
  - `MBEDTLS_CHACHA20_C`.
  and algorithms/modes:
  - `MBEDTLS_CBC_C`,
  - `MBEDTLS_CFB_C`,
  - `MBEDTLS_CTR_C`,
  - `MBEDTLS_OFB_C`,
  - `MBEDTLS_XTS_C`,
  - `MBEDTLS_CCM_C`,
  - `MBEDTLS_GCM_C`,
  - `MBEDTLS_CHACHAPOLY_C`,
  - `MBEDTLS_NULL_CIPHER`.

Once a key type and related algorithm are accelerated, all the PSA Crypto APIs
will work, as well as X.509 and TLS (with `MBEDTLS_USE_PSA_CRYPTO` enabled) but
some non-PSA APIs will be absent or have reduced functionality, see
[Restrictions](#restrictions) for details.

### Restrictions

- If an algorithm other than CCM and GCM (see
  ["Partial acceleration for CCM/GCM"](#partial-acceleration-for-ccmgcm) below)
  is enabled but not accelerated, then all key types that can be used with it
  will need to be built-in.

> - If a key type is enabled but not accelerated, then all algorithms than can be

Reviewer: ```suggestion
- If a key type is enabled but not accelerated, then all algorithms that can be
```

algorithm/mode you can:
- enable desired PSA key type(s):
  - `PSA_WANT_KEY_TYPE_AES`,
  - `PSA_WANT_KEY_TYPE_ARIA`,
  - `PSA_WANT_KEY_TYPE_CAMELLIA`,
  - `PSA_WANT_KEY_TYPE_CHACHA20`,
  - `PSA_WANT_KEY_TYPE_DES`;
- enable desired PSA algorithm(s):
  - unauthenticated ciphers modes:
    - `PSA_WANT_ALG_CBC_NO_PADDING`,
    - `PSA_WANT_ALG_CBC_PKCS7`,
    - `PSA_WANT_ALG_CCM_STAR_NO_TAG`,
    - `PSA_WANT_ALG_CFB`,
    - `PSA_WANT_ALG_CTR`,
    - `PSA_WANT_ALG_ECB_NO_PADDING`,
    - `PSA_WANT_ALG_OFB`,
    - `PSA_WANT_ALG_STREAM_CIPHER`;
  - AEADs:
    - `PSA_WANT_ALG_CCM`,
    - `PSA_WANT_ALG_GCM`,
    - `PSA_WANT_ALG_CHACHA20_POLY1305`;
- enable `MBEDTLS_PSA_ACCEL_[KEY_TYPE_xxx|ALG_yyy]` symbol(s) which correspond
   to the PSA_WANT_KEY_TYPE_xxx` and `PSA_WANT_ALG_yyy` of the previous steps;
- disable builtin support of key types:
  - `MBEDTLS_AES_C`,
  - `MBEDTLS_ARIA_C`,
  - `MBEDTLS_CAMELLIA_C`,
  - `MBEDTLS_DES_C`,
  - `MBEDTLS_CHACHA20_C`;
  and algorithms/modes:
  - `MBEDTLS_CBC_C`
  - `MBEDTLS_CFB_C`
  - `MBEDTLS_CTR_C`
  - `MBEDTLS_OFB_C`
  - `MBEDTLS_XTS_C`
  - `MBEDTLS_CCM_C`
  - `MBEDTLS_GCM_C`
  - `MBEDTLS_CHACHAPOLY_C`
  - `MBEDTLS_NULL_CIPHER`

> 

Reviewer 2: I think we're missing two important restrictions:
- if an algorithm (other than GCM and CCM, see "partial acceleration" below) is enabled but not accelerated, then all key types than can be used with it will need to be built-in;
- if a key type is enabled but not accelerated, then all algorithms than can be used with it will need to be built-in.

Note: if the future we want to lift the first restriction in general, for now it's only lifted for GCM and CCM as those are the most important ones.

(I don't think the 2nd restriction can really be lifted btw.)

  `PSA_WANT_ALG_CCM`, respectively.

### Partial acceleration for CCM/GCM

[This section depends on #8598 so it might updated while that PR progresses.]

In case legacy CCM/GCM algorithms are enabled it is still possible to benefit
from PSA acceleration by enabling support for ECB mode
(`PSA_WANT_ALG_ECB_NO_PADDING`) together with desired key type(s)
(`PSA_WANT_KEY_TYPE_[AES|ARIA|CAMELLIA]`). In such configuration it is possible
to:
- still benefit from legacy functions belonging to CCM/GCM modules
  (`mbedtls_[ccm|gcm]_xxx()`),
- disable legacy key types (`MBEDTLS_[AES|ARIA|CAMELLIA]_C`) if there is no
  other dependency requiring them, of course.

ChaChaPoly has not such feature, so it requires full acceleration (key type +
algorithm) in order to work with a driver.

### CTR-DRBG

Legacy CTR-DRBG module (enabled by `MBEDTLS_CTR_DRBG_C`) can also benefit from
PSA acceleration when:
- legacy AES module is not enabled (`MBEDTLS_AES_C`) and
- AES is supported on PSA side together with ECB mode, i.e.
  `PSA_WANT_KEY_TYPE_AES` + `PSA_WANT_ALG_ECB_NO_PADDING`.

### Disabling CIPHER_C

This is possible when:

- all ciphers and AEADs are accelerated;

> - no legacy module, either cipher or AEAD, is enabled. The only exception being

Author: Oh, right, I rephrased this section a few times in order to compact it as much a possible, but I ended up writing something not complete :/

  - `MBEDTLS_CHACHAPOLY_C`
  - `MBEDTLS_NULL_CIPHER`

Once a key type and related algorithm are accelerated, all the PSA Crypto APIs
will work, as well as X.509 and TLS (with MBEDTLS_USE_PSA_CRYPTO enabled) but
some non-PSA APIs will be absent or have reduced functionality, see

> [Disabling CIPHER_C](#disabling-cipher_c) for details.

Reviewer 2: I don't think that's about disabling `CIPHER_C`. Even if we don't disable `CIPHER_C` we don't be getting PEM decryption for example, and I don't think we'll be getting en/decryption support from PKCS(5|12) either.

At this point I'm very tempted to request a new component with `CIPHER_C` not disabled in the accelerated component (we can re-use the same reference component) just to make sure we really understand what the differences are between disabling `CIPHER_C` and leaving it enabled.

Reviewer 2: Actually if we accelerate (and disable the built-ins) for all ciphers, then even if we leave `CIPHER_C` enabled it will be pretty useless. So, the only case where leaving `CIPHER_C` enabled really makes sense is if you have some ciphers driver-only and some of them built-in. Perhaps we should have a component testing that? (But then that's probably out of scope for this PR.)

I'll note a pattern here: for ECC we also forgot to test with some curves accelerated and others not, and got some surprises when we tried it out. Here we also forgot about testing some ciphers accelerated an others not until the end. Sigh.

Author: > I'll note a pattern here: for ECC we also forgot to test with some curves accelerated and others not, and got some surprises when we tried it out. Here we also forgot about testing some ciphers accelerated an others not until the end. Sigh.

Do you think it's worth to add a new task to the ciphers/AEADs EPIC?

Reviewer 2: Yes, I think we should. I need to think a bit about exactly what it is that we want to test.

Reviewer 2: Please still fix the link: see "Restrictions" below for details. We need to clearly distinguish:
- What's related to driver-only (regardless of whether `CIPHER_C` is enabled) which goes in the "Restrictions" section.
- What's related to removing `CIPHER_C` (which can happen even without drivers), which goes in the section about that.

  `MBEDTLS_CHACHA20_C` and algorithm `MBEDTLS_POLY1305_C`.

It should be noticed that disabling `MBEDTLS_CIPHER_C` helps to reduce the
code's footprint, but unfortunately it makes the following features unavailable:
- encryption/decryption in PKCS5 and PKCS12 modules (key derivations will still
  be available),

> - encrypted PEM (write and unecrypted read work normally),

Reviewer 2: I don't think that has anything to do with `CIPHER_C`, as `pem.c` does not use `mbedtls_cipher_xxx` functions.
